Dashboard Resource Works with Grafana 8

trotttrotttrott opened the pull request:
[Grafana 8 removed `slug` endpoints](https://github.com/grafana/grafana/blob/main/CHANGELOG.md#breaking-changes) that have been deprecated since Grafana 5. My proposed fix is to:

* Deprecate the `slug` field in the provider.
* Add `uid` and use it as we were using `slug`.
* Use the read and delete endpoints that have replaced the slug endpoints.

Other changes:

* Drone will run acceptance tests with Grafana 8.0.2.
* Test configurations were moved to `examples/`. This is a convention I'd like to propose we adopt for all acceptance test configuration files. Moving them into dedicated `.tf` files allows us to take advantage of `terrform fmt`, syntax highlighting, and use them in documentation. I've prefixed these particular examples with `_acc_` to indicate they are used in acceptance tests only because they would not make good examples for documentation.
* `version` was added as a computed (read only) field. This field is removed if set as a static value in `config_json`. It cannot be managed as a static value. Therefore we manage it in state and set it on read.
* `TestAccDashboard_disappear` removed. This does not seem useful.
* 404 on read and delete fixed. The error message read as if it were removing it from state, but instead it was throwing, `Error: status: 404, body: {"message":"Dashboard not found"}` without removing it. It now displays a warning error stating that Terraform will attempt to recreate it.
* `func prepareDashboardModel(configJSON string) map[string]interface{}` replaced with `func makeDashboard(d *schema.ResourceData) gapi.Dashboard`. This function name and signature is consistent with other resources in this provider.

Closes https://github.com/grafana/terraform-provider-grafana/issues/212

trotttrotttrott committed:
Drone runs acceptances tests with Grafana 8.0.2

trotttrotttrott committed:
Move dashboard acc test configs to examples

trotttrotttrott committed:
testAccExample func

Helps read example files to be used in acceptance tests.

trotttrotttrott committed:
grafana_dashboard resource compatible w/ Grafana 8

trotttrotttrott committed:
go generate - docs generation

trotttrotttrott committed:
Fix 404 on dashboard read

trotttrotttrott committed:
Set GF_DATABASE_URL in `grafana` Drone service

julienduchesne commented on grafana/provider_test.go:
```diff
@@ -74,3 +73,14 @@ func testAccPreCheck(t *testing.T) {
+
+// testAccExample returns an example config from the examples directory.
+// Examples are used for both documentation and acceptance tests.
+func testAccExample(t *testing.T, path string) string {
```
I like this. Much better than defining hcl into strings

julienduchesne commented on grafana/resource_dashboard.go:
```diff
@@ -68,39 +86,32 @@ func ResourceDashboard() *schema.Resource {
-	slug := d.Id()
-
-	dashboard, err := client.Dashboard(slug)
+	uid := d.Id()
```
Won't the ID change make this fail when migrating from the old version to the new one? The ID stored in the state is the slug, it will try to `DashboardByUID` with a slug, which won't work, and then it will remove all dashboards from state since it won't find them

I think it might be nice to have an actual tfstate from a version before this one and use that to have a migration acceptance test

trotttrotttrott replied:
Darn, you're right. I thought this seemed too easy. I'm able to reproduce this. I'm going to see if I can add a [StateUpgrader](https://www.terraform.io/docs/extend/resources/state-migration.html).

I'm not thrilled about committing raw state for acceptance tests. Seems to me we'll only need such tests to ensure _this_ PR works. I think I'll start by documenting in fine detail how I've tested once I get a StateUpgrader working. We can decide from there if it makes sense to commit something we should maintain beyond this.

inkel commented on grafana/provider_test.go:
```diff
@@ -74,3 +73,14 @@ func testAccPreCheck(t *testing.T) {
+// Examples are used for both documentation and acceptance tests.
+func testAccExample(t *testing.T, path string) string {
+	path = fmt.Sprintf("../examples/%s", path)
+	example, err := ioutil.ReadFile(path)
```
Nit: 
```suggestion
	example, err := ioutil.ReadFile("../examples/" + path)
```

There's no actual need for `fmt.Sprintf` here, and it will definitely be much faster:

```
BenchmarkCreateSimpleString/Sprintf-12         	14443981	        79.73 ns/op	      24 B/op	       1 allocs/op
BenchmarkCreateSimpleString/plus-12            	1000000000	         0.6006 ns/op	       0 B/op	       0 allocs/op
```

inkel reviewed (approved):
Love the changes, thanks!

trotttrotttrott committed:
Rm unnecessary use of fmt.Sprintf

Co-authored-by: Leandro López <leandro.lopez@grafana.com>

trotttrotttrott committed:
StateUpgrader to migrate dashboard res from slug to uid

trotttrotttrott committed:
Cleanup

* validateDashboardConfigJSON, normalizeDashboardConfigJSON begin with
  lowercase character. There's no need to suggest these are exported.
  Same goes for basically every other function though.
* Spell unmarshal correctly in func unmarshalDashboardJSON
* Improve panic message in unmarshalDashboardJSON

trotttrotttrott committed:
Update Grafana API client

trotttrotttrott commented:
Per [@julienduchesne's comment](https://github.com/grafana/terraform-provider-grafana/pull/217#discussion_r653437169), I've added a StateUpgrader to make a smooth transition from slug to UID.

<details>
<summary>I used the following configuration to test this:</summary>

```hcl
terraform {
  required_providers {
    grafana = {
      source = "grafana/grafana"
    }
  }
}

provider "grafana" {
  url  = "http://localhost:3001"
  auth = "admin:admin"
}

resource "grafana_dashboard" "defines_uid" {
  config_json = <<EOD
{
  "title": "Defines UID",
  "uid": "defined-uid"
}
EOD
}

resource "grafana_dashboard" "no_uid" {
  config_json = <<EOD
{
  "title": "No UID"
}
EOD
}

resource "grafana_dashboard" "defines_slug" {
  config_json = <<EOD
{
  "title": "Defines Slug",
  "slug": "defined-slug"
}
EOD
}

resource "grafana_dashboard" "defines_slug_and_uid" {
  config_json = <<EOD
{
  "title": "Defines Slug and UID",
  "slug": "defined-slug",
  "uid": "defined-uid-in-addition-to-slug"
}
EOD
}
```

</details>

I first applied this configuration using the 1.11.0.

<details>
<summary>Execution plan output without the StateUpgrader produced:</summary>

```
grafana_dashboard.defines_slug_and_uid: Refreshing state... [id=defines-slug-and-uid]
grafana_dashboard.defines_uid: Refreshing state... [id=defines-uid]
grafana_dashboard.defines_slug: Refreshing state... [id=defines-slug]
grafana_dashboard.no_uid: Refreshing state... [id=no-uid]

Note: Objects have changed outside of Terraform

Terraform detected the following changes made outside of Terraform since the last "terraform apply":

  # grafana_dashboard.no_uid has been deleted
  - resource "grafana_dashboard" "no_uid" {
      - config_json  = jsonencode(
            {
              - title = "No UID"
            }
        ) -> null
      - dashboard_id = 1 -> null
      - folder       = 0 -> null
      - id           = "no-uid" -> null
      - slug         = "no-uid" -> null
    }
  # grafana_dashboard.defines_slug has been deleted
  - resource "grafana_dashboard" "defines_slug" {
      - config_json  = jsonencode(
            {
              - slug  = "defined-slug"
              - title = "Defines Slug"
            }
        ) -> null
      - dashboard_id = 4 -> null
      - folder       = 0 -> null
      - id           = "defines-slug" -> null
      - slug         = "defines-slug" -> null
    }
  # grafana_dashboard.defines_slug_and_uid has been deleted
  - resource "grafana_dashboard" "defines_slug_and_uid" {
      - config_json  = jsonencode(
            {
              - slug  = "defined-slug"
              - title = "Defines Slug and UID"
            }
        ) -> null
      - dashboard_id = 2 -> null
      - folder       = 0 -> null
      - id           = "defines-slug-and-uid" -> null
      - slug         = "defines-slug-and-uid" -> null
    }
  # grafana_dashboard.defines_uid has been deleted
  - resource "grafana_dashboard" "defines_uid" {
      - config_json  = jsonencode(
            {
              - title = "Defines UID"
            }
        ) -> null
      - dashboard_id = 3 -> null
      - folder       = 0 -> null
      - id           = "defines-uid" -> null
      - slug         = "defines-uid" -> null
    }

Unless you have made equivalent changes to your configuration, or ignored the relevant attributes using ignore_changes, the following plan may include actions to
undo or respond to these changes.

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # grafana_dashboard.defines_slug will be created
  + resource "grafana_dashboard" "defines_slug" {
      + config_json  = jsonencode(
            {
              + slug  = "defined-slug"
              + title = "Defines Slug"
            }
        )
      + dashboard_id = (known after apply)
      + id           = (known after apply)
      + slug         = (known after apply)
      + uid          = (known after apply)
      + version      = (known after apply)
    }

  # grafana_dashboard.defines_slug_and_uid will be created
  + resource "grafana_dashboard" "defines_slug_and_uid" {
      + config_json  = jsonencode(
            {
              + slug  = "defined-slug"
              + title = "Defines Slug and UID"
              + uid   = "defined-uid-in-addition-to-slug"
            }
        )
      + dashboard_id = (known after apply)
      + id           = (known after apply)
      + slug         = (known after apply)
      + uid          = (known after apply)
      + version      = (known after apply)
    }

  # grafana_dashboard.defines_uid will be created
  + resource "grafana_dashboard" "defines_uid" {
      + config_json  = jsonencode(
            {
              + title = "Defines UID"
              + uid   = "defined-uid"
            }
        )
      + dashboard_id = (known after apply)
      + id           = (known after apply)
      + slug         = (known after apply)
      + uid          = (known after apply)
      + version      = (known after apply)
    }

  # grafana_dashboard.no_uid will be created
  + resource "grafana_dashboard" "no_uid" {
      + config_json  = jsonencode(
            {
              + title = "No UID"
            }
        )
      + dashboard_id = (known after apply)
      + id           = (known after apply)
      + slug         = (known after apply)
      + uid          = (known after apply)
      + version      = (known after apply)
    }

Plan: 4 to add, 0 to change, 0 to destroy.
╷
│ Warning: Dashboard "defines-uid" is in state, but no longer exists in grafana
│
│   with grafana_dashboard.defines_uid,
│   on main.tf line 14, in resource "grafana_dashboard" "defines_uid":
│   14: resource "grafana_dashboard" "defines_uid" {
│
│ "defines-uid" will be recreated when you apply
╵
╷
│ Warning: Dashboard "no-uid" is in state, but no longer exists in grafana
│
│   with grafana_dashboard.no_uid,
│   on main.tf line 23, in resource "grafana_dashboard" "no_uid":
│   23: resource "grafana_dashboard" "no_uid" {
│
│ "no-uid" will be recreated when you apply
╵
╷
│ Warning: Dashboard "defines-slug" is in state, but no longer exists in grafana
│
│   with grafana_dashboard.defines_slug,
│   on main.tf line 31, in resource "grafana_dashboard" "defines_slug":
│   31: resource "grafana_dashboard" "defines_slug" {
│
│ "defines-slug" will be recreated when you apply
╵
╷
│ Warning: Dashboard "defines-slug-and-uid" is in state, but no longer exists in grafana
│
│   with grafana_dashboard.defines_slug_and_uid,
│   on main.tf line 40, in resource "grafana_dashboard" "defines_slug_and_uid":
│   40: resource "grafana_dashboard" "defines_slug_and_uid" {
│
│ "defines-slug-and-uid" will be recreated when you apply
╵

─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

</details>

The execution plan with the StateUpgrader shows no changes.

trotttrotttrott committed:
Use DashboardByUID testAccDashboardFolderCheckDestroy

trotttrotttrott committed:
No need for descriptions in dashboard v0 schema

trotttrotttrott committed:
V0 StateUpgrader can handle missing dashboard

trotttrotttrott committed:
Ref master of grafana-api-golang-client in go module config

trotttrotttrott committed:
V0 StateUpgrader does not return error if version cannot be set

trotttrotttrott committed:
Drone tests against Grafana 8.0.3
